### DISCUSSION TITLE
A few questions for the miniapps: Navier

### DISCUSSION THREAD

**Discussion by Watchmojo** [mfem/mfem#2011] [Q&A] [answered]
Hi, MFEM team

I am very interested at the miniapps `Navier'. If possible, I would like to further explore MFEM in the context of CFD in the future, especially DNS, LES, instability analysis, etc. 

I have a few `very stupid' questions about Navier:
1.  In navier_3dfoc.cpp, I tried to change the order of accuracy by revising ctx.order = 6, and the program gave an error: Verification failed: (D1D <= T_MAX_D1D && Q1D <= T_MAX_Q1D) is false. How to change the order of functional space? 

2. In void NavierSolver::Setup(double dt), const IntegrationRule &ir_ni = rules_ni.Get(vfes->GetFE(0)->GetGeomType(),  **2 * order -  #1**); Why is IntegrationRule ir_ni of order '2 *order-1', instead of 'order'?

3. In void NavierSolver::Setup(double dt),  There is @TODO: This order should always be the same as the nonlinear forms one! Since this is not implemented yet, is it going to affect the results? 

4. For the nonlinear term, is the discontinuity at the interface of element taken care of? 

5. For a large-scale CFD simulation, the possibility of saving and restarting from any timestep is essential. Is it difficult to implement? 

6. Is it possible to also include an example of a DNS of low-Reynolds number turbulent plane channel flow (Kim, Moin and Moser, Turbulence statistics in fully developed channel flow at low Reynolds number, JFM, 1987)? Is some stabilization approach necessary for such a low Reynolds number? 

Thank you very much.
XF

**Comment by jandrej** [comment, marked as answer]
1. This is a problem that arises with how MFEM handles the so called "partial assembly" algorithm. We allocate memory for the number of degrees of freedom in 1 dimension. But this number has to be known at compilation time. You can just change `T_MAX_D1D ` and `T_MAX_Q1D` in https://github.com/mfem/mfem/blob/24cef871af497a2ffaa9905a2253649ba3ca0f25/fem/nonlininteg_vectorconvection.cpp#L807 I know this is not an optimal solution.

2. The `NI` IntegrationRule was experimental. It is not used currently unless you call `NavierSolver::EnableNI(true)`. The order resembles co-located Gauss-Lobatto quadrature points and dofs.

3. I was not sure when I first implemented Navier. While this is not the case for the acceleration terms, I stumbled upon a test case last week where this _can_ cause to loose one order of convergence in time for accuracy because of the boundary terms. This will be fixed with #2002 but as an easy workaround, you can do:
```
BoundaryNormalLFIntegrator(*vel_dbc.coeff, 2, 2)
BoundaryNormalLFIntegrator(*FText_gfcoeff, 2, 2)
```

4. I'm not sure what you mean by that? We are using an H1 discretization here and those elements do not support discontinuities.

5. One of the benefits of Navier is that the code can be "bootstrapped" in time. So Navier does not have to implement restarts (but we can provide an example of course). The user code can decide how restarts are implemented by saving the last N solutions at the appropriate time and corresponding `dt` values. Then you can call Navier as if it would "start from zero" with a provisional time step. See [here](https://github.com/mfem/mfem/blob/master/miniapps/navier/navier_solver.hpp#L162) and an example of how it is used in the [variable stepsize example](https://github.com/mfem/mfem/blob/master/miniapps/navier/navier_kovasznay_vs.cpp). You can use any file format that mfem can read and write for the grid functions

6. I don't see a reason why this could not be implemented.

**Reply by Watchmojo** [reply]
Thank you for the prompt reply. 

--4. Please correct me if I am wrong: H1 does not guarantee the continuity of derivatives at the element interface. In navier_tgv.cpp, ComputeQCriterion: Q is averaged at the element interface, as I expected. (1) Is such averaging implemented for the nonlinear term in navier_solver as well? (2) line 207 in navier_tgv.cpp,  q(i) /= nz, is it better to weight q(i) by its inverse Jacobian? 

--6. I am very much looking forward to this, and please let me know if my input is needed.

**Reply by jandrej** [reply]
In general, averaging is applied where it is necessary. See e.g. `NavierSolver::ComputeCurlCurl2D`. There is no averaging necessary (not that I am aware of) for the nonlinear extrapolation. It is corrected in the Helmholtz step.

Let me clarify that I am not planning to implement the above mentioned example. I offer help in doing so. The algorithm benefits from Reynolds numbers >> 1. Although test examples work just fine with small Reynolds numbers ~10.

**Reply by Watchmojo** [reply]
Thank you for your reply and clarification. 

Is it possible to create 2 or even arbitrary numbers of navier_solver in one program? More importantly, each MPI communicator is designated to solve one navier_solver. For example, cores 1 to 8 solve the first navier_solver, and cores 9 to 16 solve the second. 
If this is possible, it will be very useful for simulations where one direction uses Fourier series and the other two directions uses FEM.

**Comment by goinea** [comment]
@jandrej Is aid with the above example offer still open? If so, I have a few questions. Also, would this need to be a separate issue or just continued  here?

**Reply by jandrej** [reply]
Please open a new discussion with your topic. Thank you.
